File: ROS2/src/my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool
 
 
class NumberCounterNode(Node):
    def __init__(self):
        super().__init__("number_counter")

        self.counter_ = 0
        # Create subscriber
        self.subscriber_ = self.create_subscription(Int64, "number",
                                                    self.callback_number_counter,
                                                    10)
        self.publisher_ = self.create_publisher(Int64, "number_count", 10)

        # Reset counter service
        self.server_ = self.create_service(SetBool, "reset_counter",
                                           self.callback_reset_counter)
                                                
        self.get_logger().info("Number counter has started.")

    def callback_number_counter(self, msg):
        # Subscriber
        self.counter_ += msg.data
        self.get_logger().info("Number count is: " + str(self.counter_))

        # Publisher
        new_msg = Int64()
        new_msg.data = self.counter_
        self.publisher_.publish(new_msg)

    def callback_reset_counter(self, request, response):
        # Service
        if request.data:
            self.counter_ = 0
            response.success = True
            response.message = "Counter has been reset"
        else:
            response.success = False
            response.message = "Counter has not been reset"

        return response


        bool.success


def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode()
    rclpy.spin(node)
    rclpy.shutdown()
 
 
if __name__ == "__main__":
    main()

File: ROS2/src/my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"
 
using std::placeholders::_1;
using std::placeholders::_2;

class NumberCounterNode : public rclcpp::Node
{
public:
    NumberCounterNode() : Node("number_counter"), counter_(0)
    {
        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number_count", 10);
        subscriber_ = this->create_subscription<example_interfaces::msg::Int64>("number", 10,
                        std::bind(&NumberCounterNode::callbackNumberCounter, this, std::placeholders::_1));
        // Reset counter service
        server_ = this->create_service<example_interfaces::srv::SetBool>(
                "reset_counter",
                std::bind(&NumberCounterNode::callbackResetCounter, this, _1, _2));
        RCLCPP_INFO(this->get_logger(), "Number Counter node has started.");
    }
 
private:
    void callbackNumberCounter(const example_interfaces::msg::Int64::SharedPtr msg)
    {
        // Subscriber part
        counter_ += msg->data;
        RCLCPP_INFO(this->get_logger(), "%d", counter_);
        // Publisher Part
        auto new_msg = example_interfaces::msg::Int64();
        new_msg.data = counter_;
        publisher_->publish(new_msg);
    }

    void callbackResetCounter(const example_interfaces::srv::SetBool::Request::SharedPtr request,
                              const example_interfaces::srv::SetBool::Response::SharedPtr response)
    {
        if (request->data)
        {
            counter_ = 0;
            response->success = true;
            response->message = "Counter has been reset";
        }
        else
        {
            response->success = false;
            response->message = "Counter has not been reset";
        }
    }
    int counter_;
    // Declare subscriber
    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
    // Declare Publisher
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    // Declare Service
    rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr server_;
};
 
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
